style.md

```diff
@@ -1515,6 +1516,118 @@ Note that the compiler will not generate errors when using predeclared
 identifiers, but tools such as `go vet` should correctly point out these and
 other cases of shadowing.
 
+### Avoid `init()`
+
+Avoid `init()` where possible. At a minimum, all code within `init()` should:
+
+1. Be completely deterministic, regardless of program environment or invocation.
+2. Be completely self-contained within the program.
+3. Not depend on side-effects of other `init()` functions.
+4. Be able to run in any order. While `init()` ordering is well-known, code can
+   change, and thus `init()` functions that rely on specific sequencing make
+   code brittle and error-prone.
+5. Not access or manipulate global or environment state, such as machine
+   information, environment variables, working directory, program
+   arguments/inputs, etc.
+6. Not perform I/O, including both filesystem, network, and system calls.
+
+Code that cannot satisfy these requirements likely belongs as a helper to be
+called as part of `main()` (or elsewhere in a program's lifecycle), or be
+written as part of `main()` itself. In particular, libraries that are intended
+to be used by other programs should take special care to be completely
+deterministic and not perform "init magic".
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Foo struct {
+    // ...
+}
+
+var _defaultFoo Foo
+
+func init() {
+    _defaultFoo = Foo{
+        // ...
+    }
+}
+```
+
+</td><td>
+
+```go
+var _defaultFoo = Foo{
+    // ...
+}
+```
+
+</td></tr>
+<tr><td>
+
+```go
+type Config struct {
+    // ...
+}
+
+var _config Config
+
+func init() {
+    // Bad: based on current directory
+    cwd, _ := os.Getwd()
+
+    // Bad: I/O
+    raw, _ := ioutil.ReadFile(
+        path.Join(cwd, "config", "config.yaml"),
+    )
+
+    yaml.Unmarshal(raw, &_config)
+}
+```
+
+</td><td>
+
+```go
+type Config struct {
+    // ...
+}
+
+func loadConfig() Config {
+    cwd, _ := os.Getwd()
+    raw, _ := ioutil.ReadFile(
+        path.Join(cwd, "config", "config.yaml"),
+    )
+
+    var config Config
+    yaml.Unmarshal(raw, &config)
+
+    return config
+}
+```
+
+</td></tr>
+</tbody></table>
+
+Notable exceptions to `init()` determinism guidance include situations such as
+seeding non-cryptographic random number generators used across a codebase:
+
+```go
+func init() {
+    rand.Seed(time.Now().UnixNano())
+}
+```
+
+Considering the above, some situations in which `init()` may be preferable or
+necessary might include:
+
+- To reduce or prevent user error
```

Comment:
This is unclear

Reply:
Yeah, I didn't really have a good concrete example offhand other than something about providing sane defaults (like in situations where folks don't control the zero values or there isn't a useful zero value). Could go either way on this one, really.

Reply:
We don't necessarily need examples for all exceptions, but for this one, it's
not clear to me how init would help here. We can drop it and add it back later
if it becomes more obvious with different phrasing/with an example.